File: nodejs/app_rest.js
const express = require('express');
const bodyParser = require('body-parser');

const app = express();
const port = 3000;

app.use(bodyParser.json());

let tarefas = [
    { id: 1, titulo: 'Estudar Node.js', feito: false },
    { id: 2, titulo: 'Fazer compras', feito: false }
];

app.get('/tarefas', (req, res) => {
    res.json({ tarefas });
});

app.post('/tarefas', (req, res) => {
    try {
        const novaTarefa = { id: tarefas.length + 1, titulo: req.body.titulo, feito: false };
        tarefas.push(novaTarefa);
        res.status(201).json({ tarefa_adicionada: novaTarefa });
    } catch (error) {
        res.status(400).json({ error: 'Titulo da tarefa é obrigatório.' });
    }
});

app.get('/tarefas/:tarefaId', (req, res) => {
    const tarefa = tarefas.find(tarefa => tarefa.id === parseInt(req.params.tarefaId));
    if (tarefa) {
        res.json({ tarefa });
    } else {
        res.status(404).json({ message: 'Tarefa não encontrada' });
    }
});

app.put('/tarefas/:tarefaId', (req, res) => {
    const tarefa = tarefas.find(tarefa => tarefa.id === parseInt(req.params.tarefaId));
    if (tarefa) {
        tarefa.titulo = req.body.titulo || tarefa.titulo;
        tarefa.feito = req.body.feito || tarefa.feito;
        res.json({ tarefa_updated: tarefa });
    } else {
        res.status(404).json({ message: 'Tarefa não encontrada' });
    }
});

app.delete('/tarefas/:tarefaId', (req, res) => {
    tarefas = tarefas.filter(tarefa => tarefa.id !== parseInt(req.params.tarefaId));
    res.json({ message: 'Tarefa deletada' });
});

app.listen(port, () => {
    console.log(`Servidor rodando em http://localhost:${port}`);
});

File: golang_rest/app_rest.go
package main

import (
	"encoding/json"
	"log"
	"net/http"
	"strconv"

	//go get -u github.com/gorilla/mux
	"github.com/gorilla/mux"
)

type Tarefa struct {
	ID     int    `json:"id"`
	Titulo string `json:"titulo"`
	Feito  bool   `json:"feito"`
}

var tarefas = []Tarefa{
	{ID: 1, Titulo: "Estudar Go", Feito: false},
	{ID: 2, Titulo: "Fazer compras", Feito: false},
}

func pegaTarefa(w http.ResponseWriter, r *http.Request) {
	json.NewEncoder(w).Encode(tarefas)
}

func criaTarefa(w http.ResponseWriter, r *http.Request) {
	var newtarefa Tarefa
	_ = json.NewDecoder(r.Body).Decode(&newtarefa)
	newtarefa.ID = len(tarefas) + 1
	tarefas = append(tarefas, newtarefa)
	w.WriteHeader(http.StatusCreated)
	json.NewEncoder(w).Encode(newtarefa)
}

func pegaTarefas(w http.ResponseWriter, r *http.Request) {
	params := mux.Vars(r)
	tarefaID, _ := strconv.Atoi(params["tarefaID"])
	for _, tarefa := range tarefas {
		if tarefa.ID == tarefaID {
			json.NewEncoder(w).Encode(tarefa)
			return
		}
	}
	json.NewEncoder(w).Encode(map[string]string{"message": "Tarefa não encontrada"})
}

func atualizaTarefa(w http.ResponseWriter, r *http.Request) {
	params := mux.Vars(r)
	tarefaID, _ := strconv.Atoi(params["tarefaID"])

	for index, tarefa := range tarefas {
		if tarefa.ID == tarefaID {
			var updatedtarefa Tarefa
			_ = json.NewDecoder(r.Body).Decode(&updatedtarefa)
			tarefas[index].Titulo = updatedtarefa.Titulo
			tarefas[index].Feito = updatedtarefa.Feito
			json.NewEncoder(w).Encode(tarefas[index])
			return
		}
	}
	json.NewEncoder(w).Encode(map[string]string{"message": "Tarefa não encontrada"})
}

func apagaTarefa(w http.ResponseWriter, r *http.Request) {
	params := mux.Vars(r)
	tarefaID, _ := strconv.Atoi(params["tarefaID"])

	for index, tarefa := range tarefas {
		if tarefa.ID == tarefaID {
			tarefas = append(tarefas[:index], tarefas[index+1:]...)
			json.NewEncoder(w).Encode(map[string]string{"message": "Tarefa deletada"})
			return
		}
	}
	json.NewEncoder(w).Encode(map[string]string{"message": "Tarefa não encontrada"})
}

func main() {
	router := mux.NewRouter()

	router.HandleFunc("/tarefas", pegaTarefas).Methods("GET")
	router.HandleFunc("/tarefas", criaTarefa).Methods("POST")
	router.HandleFunc("/tarefas/{tarefaID}", pegaTarefas).Methods("GET")
	router.HandleFunc("/tarefas/{tarefaID}", atualizaTarefa).Methods("PUT")
	router.HandleFunc("/tarefas/{tarefaID}", apagaTarefa).Methods("DELETE")

	log.Fatal(http.ListenAndServe(":8080", router))
}
